### PR TITLE
rule: prevent misconfiguration of placement rules in some situations

### PR DESCRIPTION
**Pull request by disksing**
### What problem does this PR solve? 
Fix https://github.com/pingcap/pd/issues/2476

### What is changed and how it works?
when update placement rules, return error if after update:
- there is no rule lest
- there is gap between rules (region may match no rules)

### Check List 

Tests 
 - Unit test
 - Manual test (add detailed scripts or steps below)

Related changes
- Need to cherry-pick to the release branch: release-4.0

### Release note 
- Prevent misconfiguration of placement rules in some situations

### CODE REVIEW

File: server/schedule/placement/rule_manager.go [hunk 1]
```diff
@@ -169,8 +173,13 @@ func (m *RuleManager) SetRule(rule *Rule) error {
...
 
-	if err = m.store.SaveRule(rule.StoreKey(), rule); err != nil {
-		// restore
+	ruleList, err := buildRuleList(m.rules)
+	if err == nil {
+		err = m.store.SaveRule(rule.StoreKey(), rule)
+	}
+
+	// restore
+	if err != nil {
```
Comment (rleungx) [inline server/schedule/placement/rule_manager.go:182]: else?

Comment (disksing) [reply server/schedule/placement/rule_manager.go:182]: cannot use `else`, `err` may become not nil in if

Comment (rleungx) [reply server/schedule/placement/rule_manager.go:182]: got it.

**Pull request merged**